lib/common/socket.c

```diff
@@ -828,6 +833,18 @@ static void on_handshake_complete(h2o_socket_t *sock, const char *err)
         }
     }
 
+    if (sock->ssl->handshake.client.session_cache) {
+        if (err == NULL || err == h2o_socket_error_ssl_cert_name_mismatch) {
+            SSL_SESSION *session = SSL_get1_session(sock->ssl->ssl);
+            h2o_cache_set(sock->ssl->handshake.client.session_cache, h2o_now(h2o_socket_get_loop(sock)),
+                          sock->ssl->handshake.client.session_cache_key, sock->ssl->handshake.client.session_cache_key_hash,
+                          h2o_iovec_init(session, 1));
+        } else {
+            h2o_cache_delete(sock->ssl->handshake.client.session_cache, h2o_now(h2o_socket_get_loop(sock)),
+                             sock->ssl->handshake.client.session_cache_key, sock->ssl->handshake.client.session_cache_key_hash);
```

Comment:
Is there a need to remove a session cache upon failure?

The reason I ask this is there could be other reasons than the server loosing necessary information to resume a session that could cause the handshake to fail (e.g. TCP connection reset), in which case we _might_ not want to remove the cache entry.